feat: add cluster.duplicate() method

SGrondin opened the pull request:
Hi,

See https://github.com/luin/ioredis/issues/921. This PR adds the missing `cluster.duplicate()` method.

Notes:
- I'm using `this.startupNodes.slice(0)` to ensure we pass a copy and not a reference. This matches the behavior of the options, which also get a copy thanks to `Object.assign()`.
- I'm not sure this new test I added is being executed when running `npm test`. I don't see it in the mocha output. It works as expected when I run ` npx mocha test/functional/cluster/duplicate.ts`, could someone please validate?

luin commented:
LGTM. The test should be executed since `npm test` simply calls `mocha test/**/*.ts`. "test/functional/cluster/duplicate.ts" matches the pattern. Does `npx mocha test/**/*.ts` execute the test?

SGrondin commented:
Wow, that's a quick response!

Unfortunately `npx mocha test/**/*.ts` fails to run any tests because of a compilation error in `test/functional/scan_stream.ts`. I'm guessing it's also why `npm test` isn't running the new test.

```
TSError: ⨯ Unable to compile TypeScript:
test/functional/scan_stream.ts:239:15 - error TS2339: Property 'sadd' does not exist on type 'Cluster'.
test/functional/scan_stream.ts:240:30 - error TS2339: Property 'sscanStream' does not exist on type 'Cluster'.
```

luin commented:
I see. There are some issues related to type which will be fixed in the next major version. Just pushed a temp fix. Please merge the latest commit and try again.

SGrondin commented:
I've rebased on top of the latest commit and fixed the code formatting issues. The build is passing but I still don't see the new test. It should be under a `describe` grouped called `#duplicate`. All I see is the other duplicate test, for the base client.

luin commented:
It seems that mocha ignores tests that contain TypeScript errors, which is the reason the newly added test wasn't executed, as well as some other tests. Didn't notice this issue before.

luin commented:
@SGrondin Just did some fixes for the issues. Merge the latest commits will do the trick.

SGrondin commented:
Very nice. It ran 55 tests that were previously ignored.
